fl4health/clients/perfcl_client.py

```
@@ -0,0 +1,257 @@
+        but AFTER server-side aggregation.
+
+        Args:
+            current_server_round (int): Current server round being performed.
+        """
+        # Save the parameters of the aggregated global model
+        assert isinstance(self.model, PerFclModel)
+        self.initial_global_module = self.clone_and_freeze_model(self.model.second_module)
+
+        super().update_before_train(current_server_round)
+
+    def compute_loss_and_additional_losses(
+        self,
+        preds: Dict[str, torch.Tensor],
+        features: Dict[str, torch.Tensor],
+        target: torch.Tensor,
+    ) -> Tuple[torch.Tensor, Dict[str, torch.Tensor]]:
+        """
+        Computes the loss and any additional losses given predictions of the model and ground truth data.
+        For PerFCL, the total loss is the standard criterion loss provided by the user and the PerFCL contrastive
+        losses aimed at manipulating the local and global feature extractor latent spaces.
+
+        Args:
+            preds (Dict[str, torch.Tensor]): Prediction(s) of the model(s) indexed by name.
+            features (Dict[str, torch.Tensor]): Feature(s) of the model(s) indexed by name.
+            target (torch.Tensor): Ground truth data to evaluate predictions against.
+
+        Returns:
+            Tuple[torch.Tensor, Union[Dict[str, torch.Tensor], None]]; A tuple with:
+                - The tensor for the total loss
+                - A dictionary with `loss`, `total_loss`, `global_feature_contrastive_loss`, and
+                    `local_feature_contrastive_loss` representing the various and relevant pieces of the loss
+                    calculations
+        """
+
+        loss = self.criterion(preds["prediction"], target)
+        # If any of these are None then we don't compute the PerFCL loss. This will happen on the first client-side
+        # training run.
+        if self.old_local_module is None or self.old_global_module is None or self.initial_global_module is None:
```

Should we also assert its the first round and throw something if they are not provided and its not the first round. Or we can log something. Just to raise if we are not computing some of the losses and its not the first round meaning there is likely an error happening somewhere. If you think its unlikely to happen though you can disregard this

---

Talk about this together, this is a little more complicated to do but we should do it. So a ticket has been filed in the backlog: https://app.clickup.com/t/8688ptp72